File: CS/Client/BallGameClient/ClientGUI.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;

namespace BallGameClient
{
    public partial class ClientGUI : Form
    {
        private Client client;

        public ClientGUI()
        {
            InitializeComponent();
            this.CreateHandle();

            client = new Client(this);

            SetIdLabel(client.GetPlayerId());
            SetBallHolderLabel(client.GetBallHolder());

            Client.inGamePlayers = client.GetPlayers();
            foreach (int id in Client.inGamePlayers)
            {
                if (id != client.GetPlayerId())
                {
                    AddIdToGui(id);
                }
            }
        }
        [STAThread]
        static void Main()
        {
            Application.EnableVisualStyles();
            Application.SetCompatibleTextRenderingDefault(false);
            Application.Run(new ClientGUI());
        }

        private void passBtn_Click(object sender, EventArgs e)
        {
            if (client.GetPlayerId() == client.GetBallHolder())
            {
                int passBallTo;
                try
                {
                    passBallTo = int.Parse(idField.Text);
                    if (client.IsValidPlayer(passBallTo))
                    {
                        client.PassBall(int.Parse(idField.Text));
                    }
                    else
                    {
                        MessageBox.Show("That player is not in the game!");
                    }
                }
                catch (FormatException)
                {
                    MessageBox.Show("Invalid ID");
                }
            }
            else
            {
                MessageBox.Show("You do not have the ball");
            }
        }

        public void AddIdToGui(int id)
        {
            try
            {
                String idString = (id == client.GetPlayerId()) ? id + " (You)\n" : id + "\n";
                playersArea.Invoke(new Action(() => playersArea.AppendText(idString)));
            }
            catch (InvalidOperationException)
            {
                Console.WriteLine("Error adding ID to GUI");
            }
        }

        public void RemoveIdFromGui(int id)
        {
            playersArea.Invoke(new Action(() =>
            {
                string[] allIdsArr = playersArea.Text.Split(new[] { Environment.NewLine }, StringSplitOptions.None);
                List<string> allIdsList = new List<string>();
                foreach (string playerId in allIdsArr)
                {
                    Console.WriteLine($"playerId = {playerId}");
                    string newPlayerId = (playerId.Contains("You")) ? playerId.Split(' ')[0] : playerId;
                    Console.WriteLine($"newPlayerId = {newPlayerId}");
                    allIdsList.Add(newPlayerId);
                }
                allIdsList.Remove(id.ToString());
                playersArea.Invoke(new Action(() => playersArea.Text = ""));
                foreach (string playerId in allIdsList)
                {

                    AddIdToGui(int.Parse(playerId));
                }
            }));
        }

        public void AddBroadcastToGui(string broadcastMessage)
        {
            try
            {
                logArea.Invoke(new Action(() => logArea.AppendText(broadcastMessage + "\n")));
            }
            catch (InvalidOperationException e)
            {
                Console.WriteLine("Error while appending to log\n" + e.Message);
            }
        }

        public void SetIdLabel(int id)
        {
            idLabel.Invoke(new Action(() => idLabel.Text = "Your ID: " + id));
        }

        public void SetBallHolderLabel(int id)
        {
            string ballHolder = (id == client.GetPlayerId()) ? "Ball Holder: You" : "Ball Holder: " + id;
            ballHolderLabel.Invoke(new Action(() => ballHolderLabel.Text = ballHolder));
        }

        private void ClientGUI_FormClosed(object sender, FormClosedEventArgs e)
        {
            Environment.Exit(1);
        }
    }
}

File: Java/Client/src/ClientGUI.java
import javax.swing.*;
import java.awt.*;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class ClientGUI {

    private JTextField idField;
    private JButton passBtn;
    private JPanel panelMain;
    private JTextArea playersArea;
    private JTextArea logArea;
    private JLabel idLabel;
    private JLabel ballHolderLabel;
    private Client client;

    private ClientGUI() throws InterruptedException, IOException {

        logArea.setEditable(false);
        playersArea.setEditable(false);

        client = new Client(this);

        setIdLabel(client.getPlayerId());
        setBallHolderLabel(client.getBallHolder());

        Client.inGamePlayers = client.getPlayers();
        for (int id : Client.inGamePlayers) {
            if (id != client.getPlayerId()) {
                addIdToGui(id);
            }
        }

        passBtn.addActionListener(event -> {
            if (client.getPlayerId() == client.getBallHolder()) {
                int passBallTo;
                try {
                    passBallTo = Integer.parseInt(idField.getText());
                    if (client.isValidPlayer(passBallTo)) {
                        client.passBall(Integer.parseInt(idField.getText()));
                    } else {
                        JOptionPane.showMessageDialog(null, "That player is not in the game!");
                    }
                } catch (NumberFormatException e) {
                    JOptionPane.showMessageDialog(null, "Invalid ID!");
                }
            } else {
                JOptionPane.showMessageDialog(null, "You do not have the ball!");
            }
        });
    }

    public static void main(String[] args) throws InterruptedException, IOException {
        JFrame frame = new JFrame("CE303 Assignment");
        frame.setContentPane(new ClientGUI().panelMain);
        frame.setMinimumSize(new Dimension(500, 400));
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.setVisible(true);
        frame.setAlwaysOnTop(true);
    }

    void addIdToGui(int id) {
        String idString = (id == client.getPlayerId()) ? id + " (You)\n" : id + "\n";
        playersArea.append(idString);
        playersArea.repaint();
    }

    void removeIdFromGui(int id) {
        String[] allIdsArr = playersArea.getText().split("\n");
        List<String> allIdsList = new ArrayList<>();
        for (String playerId : allIdsArr) {
            playerId = (playerId.contains("You")) ? playerId.substring(0, playerId.length() - 6) : playerId;
            allIdsList.add(playerId);
        }
        allIdsList.remove(Integer.toString(id));
        playersArea.setText("");
        for (String playerId : allIdsList) {
            addIdToGui(Integer.parseInt(playerId));
        }
    }

    void addBroadcastToGui(String broadcastMessage) {
        logArea.append(broadcastMessage + "\n");
        logArea.repaint();
    }

    void setIdLabel(int id) {
        idLabel.setText("Your ID: " + id);
    }

    void setBallHolderLabel(int id) {
        String ballHolder = (id == client.getPlayerId()) ? "Ball Holder: You" : "Ball Holder: " + id;
        ballHolderLabel.setText(ballHolder);
        ballHolderLabel.repaint();
    }
}
